lib/eventHubSender.ts

```diff
@@ -0,0 +1,335 @@
+      try {
+        await rheaPromise.closeSender(this._sender);
+        this.removeAllListeners();
+        delete this._context.senders[this.name!];
+        debug(`Deleted the sender "${this.name!}" from the client cache.`);
+        this._sender = undefined;
+        this._session = undefined;
+        clearTimeout(this._tokenRenewalTimer as NodeJS.Timer);
+        debug(`[${this._context.connectionId}] Sender "${this.name}" closed.`);
+      } catch (err) {
+        return Promise.reject(err);
+      }
+    }
+  }
+
+  /**
+   * Tries to send the message to EventHub if there is enough credit to send them
+   * and the circular buffer has available space to settle the message after sending them.
+   *
+   * We have implemented a synchronous send over here. We shall be waiting for the message
+   * to be accepted or rejected and accordingly resolve or reject the promise.
+   *
+   * @param message The message to be sent to EventHub.
+   * @return {Promise<rheaPromise.Delivery>} Promise<rheaPromise.Delivery>
+   */
+  private _trySend(message: AmqpMessage, tag?: any, format?: number): Promise<rheaPromise.Delivery> {
+    return new Promise((resolve, reject) => {
+      debug(`[${this._context.connectionId}] Sender "${this.name}", credit: ${this._sender.credit}, ` +
+        `available: ${this._sender.session.outgoing.available()}.`);
+      if (this._sender.sendable()) {
+        debug(`[${this._context.connectionId}] Sender "${this.name}", sending message: \n`, message);
+        let onRejected: Func<rheaPromise.Context, void>;
+        const onAccepted: Func<rheaPromise.Context, void> = (context: rheaPromise.Context) => {
+          // Since we will be adding listener for accepted and rejected event every time
+          // we send a message, we need to remove listener for both the events.
+          // This will ensure duplicate listeners are not added for the same event.
+          this._sender.removeListener("accepted", onAccepted);
+
+          this._sender.removeListener("rejected", onRejected);
+          debug(`[${this._context.connectionId}] Sender "${this.name}", got event accepted.`);
+          resolve(context.delivery);
+        };
+        onRejected = (context: rheaPromise.Context) => {
+          this._sender.removeListener("rejected", onRejected);
+          this._sender.removeListener("accepted", onAccepted);
+          debug(`[${this._context.connectionId}] Sender "${this.name}", got event accepted.`);
+          reject(errors.translate(context!.delivery!.remote_state.error));
+        };
+        this._sender.on("accepted", onAccepted);
+        this._sender.on("rejected", onRejected);
+        const delivery = this._sender.send(message, tag, format);
+        debug(`[${this._context.connectionId}] Sender "${this.name}", sent message with delivery id: ${delivery.id}`);
+      } else {
+        // This case should technically not happen. rhea starts the sender credit with 1000 and the circular buffer with a size
```

Comment:
this is an infinite loop. it shouldn't happen under normal circumstances, but if the event hub gets behind and isn't giving your credits back or there's some other case that causes `sendable()` to be false (maybe link detached by event hub?), you'll try to send every message forever on a 5 second interval without ever resolving/rejecting. there should be a retry limit and/or a buffer limit to prevent things from hanging or the memory from running out. maybe the internal 2048 message buffer of rhea is sufficient?

there will always be cases where you can push things hard enough to max your limit out but i find it's better to inform the user (via an error) that they're pushing things too hard than to keep accepting new work until the process crashes